dask/distributed/scheduler.py

```diff
@@ -141,33 +150,42 @@ def _listen_to_workers(self):
         """ Event loop: Listen to worker router """
         while self.status != 'closed':
             try:
-                if not self.to_workers.poll(100):
+                socks = dict(self.worker_poller.poll(100))
+                if not socks:
                     continue
             except zmq.ZMQError:
                 break
-            address, header, payload = self.to_workers.recv_multipart()
-
-            header = pickle.loads(header)
-            if 'address' not in header:
-                header['address'] = address
-            log(self.address_to_workers, 'Receive job from worker', header)
-
-            try:
-                function = self.worker_functions[header['function']]
-            except KeyError:
-                log(self.address_to_workers, 'Unknown function', header)
-            else:
-                future = self.pool.apply_async(function, args=(header, payload))
+            if self.send_to_workers_recv in socks:
+                self.send_to_workers_recv.recv()
+                while not self.send_to_workers_queue.empty():
```

Comment:
I don't know if you care about starvation (or if starvation is even possible in this case), but IOLoop does this [a bit](https://github.com/tornadoweb/tornado/blob/v4.1.0/tornado/ioloop.py#L885) [differently](https://github.com/tornadoweb/tornado/blob/v4.1.0/tornado/ioloop.py#L752):

```python
with self._callback_lock:
    current_callbacks = self._callbacks
    self._callbacks = []
```

which avoids starvation that could be caused by threads rapidly putting new sends on the queue. It ensures that only the callbacks registered at the beginning of a given iteration are called during that iteration before handling incoming requests.